README.md
SWARCO Embedded Linux V2 operating system
=========================================

<img src="https://www.swarco.com/var/em_plain_site/storage/images/media/images/swarco-traffic-systems/interurban/primos/primos_compact_ohnetouch_grau/1166093-1-eng-US/Primos_Compact_ohneTouch_grau_315px.jpg" width="200" title="SWARCO CCM2200 CPU" alt="">


| :warning: Warning <br> This is a legacy SWARCO linux environment used for CCM2200 board until 2016. <br> It is only for documentation purposes, please don't use it for starting new projects.<br>See [SWARCO Linux V3](https://github.com/swarco/swarco-linux-v3) for the current environment |
|--------------------|

# Getting started

1. Install packages needed for SWARCO Linux build

    * Debian / Ubuntu Linux

        ```sh
        apt-get install git cpp gcc g++ make
        apt-get install libc6-dev libncurses5-dev
        apt-get install gettext texinfo flex bison autoconf
        apt-get install liblzo2-dev liblzo2-2 libz-dev
        ```

    * Fedora Linux

        ```sh
        yum install git autoconf automake binutils bison flex gcc gcc-c++
        yum install gdb gettext libtool texinfo make strace ncurses-devel
        yum install lzo-devel zlib-devel patch
        ```


2. Clone the repo: 

    ```sh
        git clone https://github.com/swarco/swarco-linux-v2
    ```
    
3. Download needed packages and repos and prepare our build tree (be
  prepared, this will download over 1.6G of data...):

    ```sh
        cd swarco-linux-v2
        ./prepare_tree_github.sh
    ```

4. compile everything (now you shoud grab a coffee, this can take an
  hour or more...):

    ```sh
        make
    ```

# Output of a successfully build (in directory `tftp_root`):

| File                         | Description                                                            |
|------------------------------|------------------------------------------------------------------------|
| ccm2200.bin                  | U-Boot script for updating kernel and rootfs with an USB flash drive   |
| rootfs-ccm2200-lp-nand.jffs2 | Image of JFFS2 rootfs which can be written to NAND-flash with U-Boot   |
| u-boot-ccm2200dev.bin        | Binary Image of U-Boot (for installation with JTAG or USB flash drive) |
| u-boot-update/               | Files for updating u-boot with an USB flash drive                      |
| uImage-ccm2200dev.bin        | Linux Kernel image in u-boot image format                              |


# Documentation

This repository contains a skeleton for the SWARCO Embedded Linux V2
build environment. It also contains a Makefile for controlling the overall-
build process.

The complete environment will be populated using the prepare_tree.sh.
This script extracts the the necessary source and build system files
from buildroot, Linux kernel and uBoot sources at right places and applies the
needed patches.

An old version of the documentation for SWARCO Embedded Linux V2 can be
found under:

[doc/SWARCO-LINUX.md](doc/SWARCO-LINUX.md)


# License Information 

This project contains code from Buildroot project (from
[http://git.buildroot.net/buildroot](http://git.buildroot.net/buildroot)) which is a base for the SWARCO
Embedded Linux V2 operating system used on the CCM2200 CPU.

Buildroot itself and all files in this repository contributed by
SWARCO are licensed under the
[GNU General Public License, version 2](http://www.gnu.org/licenses/old-licenses/gpl-2.0.html)
or (at your option) any later version, with the exception of the
package patches.

Buildroot also bundles patch files, which are applied to the sources
of the various packages. Those patches are not covered by the license
of Buildroot. Instead, they are covered by the license of the software
to which the patches are applied. When said software is available
under multiple licenses, the Buildroot patches are only provided under
the publicly accessible licenses.

The packages included in Buildroot and SWARCO Linux are licensed under
various open source licenses.  Some licenses require you to publish
the license text in the documentation of your product. Others require
you to redistribute the source code of the software to those that
receive your product. Please refer the Buildroot documentation and the
license documentation of each used package when distributing a product
based on this software.


Changelog
2017-06-22  Guido Classen  <gc@STSTR12221.swarco.loc>

	* buildroot/buildroot-2.0/package/netsnmp/Config.in:
	Make netsnmp MIB-modules configurable via buildroot .config
	Enable modules ucd-snmp/disk and ucd-snmp/loadave

	* buildroot/buildroot-2.0/package/openvpn/openvpn.mk:
	Security upgrade to openvpn version 2.3.17

2017-06-01  Guido Classen  <gc@STSTR12221.swarco.loc>

	* board/swarco/busybox.config:
	Upgrade busybox to git version 1.27-git-f547041

	This includes offical patch from Denys Vlasenko
	to prevent interruption of buildin read on SIGCHLD

2017-05-04  Guido Classen  <gc@STSTR12221.swarco.loc>

	* buildroot/buildroot-2.0/.config:
	Updated busybox to version 1.26.2

	* prepare_tree_github.sh (KERNEL_BASE):
	Updated Kernel to Version Tag v3.4.113-ccm2200

2017-04-21  Guido Classen  <gc@STSTR12221.swarco.loc>

	* buildroot/buildroot-2.0/swarco/Makefile:
	Advanced version number to "SWARCO Traffic Systems GmbH Embedded
	Linux v2.03"

2017-03-24  Guido Classen  <gc@STSTR12221.swarco.loc>

	* board/swarco/ccm2200/rootfs_overlay/etc/ppp/shell_gprs_dial.sh:
	Added basic support for Telit UC864-G 3G terminal adapter module

2017-03-23  Guido Classen  <gc@STSTR12221.swarco.loc>

	* board/swarco/ccm2200/rootfs_overlay/etc/ppp/shell_gprs_dial.sh:
	Added basic support for Cinterion PLS8-E LTE terminal adapter
	
        Added support for Cinterion PLS8-E with Service Interface Configuration
        2 (Modem interface over USB) using normal ppp connection.
        The direct cdc_ether network interfaces are currently NOT supported

2016-11-21  Guido Classen  <gc@STSTR12221.swarco.loc>

	* board/swarco/ccm2200/rootfs_overlay/etc/network/ip-config:
	Service IP address is changed from 169.254.17.101 to 10.99.99.99

	The IPv4 address block 169.254.0.0/16 is referred to as a "link
	local", so it can not be routed. This can cause problems on
	service connection via router (e.g., wireless router, or VPN
	router for remote access), especially on Apple iOS operating
	system.

	To avoid the problem comes the following address will used (also
	used by MicroTechnic and SWARCO Technology):

	IP address: 10.99.99.99
	Netmask: 255.255.255.248

2016-09-14  Guido Classen  <gc@STSTR12221.swarco.loc>

	* board/swarco/ccm2200/kernel/2.6.21/kernel-config-2.6.21-ccm2200 (CONFIG_BONDING):
	Enabled Network Bonding:
        - enabled ethernet bonding driver in kernel (as module) and
        - enabled ifenslave utility in busybox


2016-08-22  Guido Classen  <gc@STSTR12221.swarco.loc>
	* kernel/linux-2.6.21-ccm2200/drivers/usb/gadget/ether.c:

	added Microsot OS descriptors to RNDIS USB gadget driver.  This
	causes that Windows will automatically install the RNDIS 6.0
	driver when our device is connected. So we don't need to install
	linux.inf anymore.
    
        Also supported by Windows 10.
    
        When our device is not recognized, please remove the registry hive
	with our product/vendor id under
    
        HKEY_LOCAL_MACHINE/SYSTEM/CurrentControlSet/Control/UsbFlags
    
        this can be archived using folowing .reg registry file:
    
--------------------------- del_usb_ccm2200.reg ------------------------------
REGEDIT4
[-HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\usbflags\0525A4A20213]

---------------------------------- cut here ----------------------------------

	* kernel/linux-2.6.21-ccm2200/drivers/char/ccm2200_gpio.c:
	for compatibility with new CCMxCORE board's reset state we will
	set the digital outputs to '0', so we turn OFF the optocoupler's
	LED (and also MOSFET output)

         => This enshures an electrically safe initial state of the system
	
	* board/swarco/ccm2200/kernel/2.6.21/kernel-config-2.6.21-ccm2200:
	  Enabled kernel options:
            - network bridge (module bridge.ko)
            - IP: advanced router / policy routing / multiple routing tables

2016-05-20  Guido Classen  <gc@tux-mobil.loc>

	* buildroot/buildroot-2.0/swarco/package/tinc/tinc.mk: 
	added new package for Tinc Virtual Private Network

2016-05-12  Guido Classen  <gc@tux-mobil.loc>

	* buildroot/buildroot-2.0/ch_conf/etc/ppp/shell_gprs_dial.sh (GPRS_STATUS_FILE): 
	fixed typo error in variable GPRS_STATUS_FILE introduced by last commit

2016-05-03  Guido Classen  <gc@tux-mobil.loc>

	* buildroot/swarco_ch_conf/etc/ppp/shell_gprs_dial.sh: 
	added support for GPRS_NET_ACCESS_TYPE on Cinterion EHS5 and Huawei
	decode network status information of Huawei adapter

2016-05-02  Guido Classen  <gc@tux-mobil.loc>

	* buildroot/swarco_ch_conf/etc/ppp/shell_gprs_dial.sh: 
	Cinterion EHS5 must be provided with PDP credentials addition to
	ppp negation using AT^SGAUTH

2016-04-21  Guido Classen  <gc@tux-mobil.loc>

	* buildroot/swarco_ch_conf/etc/udhcp-usb0.conf: 
	disabled router option

	* buildroot/swarco_ch_conf/etc/ppp/shell_gprs_dial.sh: 
	close file handle, so device (e.g. /dev/ttyUSB0) can be reallocated on
	Modem Reset

	Support for Huawei E353 Stick with Product ID 1446 and 1506

2016-04-01  Guido Classen  <gc@tux-mobil.loc>

	* buildroot/swarco_ch_conf/etc/ppp/shell_gprs_dial.sh (Huawei E220, E230, E270, E870): 
	huaweiAktBbo (only for Huawei E220, E230, E270, E870) is now replaced by usb_modeswitch 
	which works for switching all devices

2016-03-31  Guido Classen  <gc@tux-mobil.loc>

	* buildroot/buildroot-2.0/package/openvpn/openvpn.mk (OPENVPN_VERSION): 
	Update openvpn to Version 2.3.10

	* buildroot/buildroot-2.0/package/busybox/Config.in (if BR2_BUSYBOX_VERSION_1_20_X):
	Update busybox to Version 1.24.2
	Provided updated config file buildroot/buildroot-2.0/swarco/busybox.config

	* buildroot/buildroot-2.0/package/stunnel/stunnel.mk (STUNNEL_SITE): 
	Update stunnel to version 5.31
	Patch to disable stack-protector (not supported by uclibc)

	* buildroot/buildroot-2.0/package/openssh/openssh.mk (OPENSSH_VERSION): 
	Update OpenSSH to Version 7.2p2
	* buildroot/buildroot-2.0/ch_conf/etc/sshd_config (PubkeyAcceptedKeyTypes  +ssh-dss): 

prepare_tree_github.sh
#!/bin/sh
#*****************************************************************************
#* 
#*  @file          prepare_tree_github.sh
#*
#*                 Unpacks packages for SWARCO Embedded Linux V2
#*                 using the respective Github repositories and
#*                 prepares Linux build tree for compilation
#*                 2010-2016 SWARCO Traffic Systems GmbH
#*
#*  @version       0.1 (\$Revision$)
#*  @author        Guido Classen
#*
#*  @par change history:
#*    2016-08-04 gc: initial version 
#*  
#*****************************************************************************

TFTP_ROOT_DIR=tftp_root
BUILDROOT_BASE=buildroot
BUILDROOT_DIR=buildroot-2.0
BUILDROOT_SOFT_FLOAT_DIR=buildroot-2.0-soft-float
U_BOOT_BASE=u-boot
U_BOOT_DIR=u-boot-v2010.09-ccm2200
KERNEL_BASE=kernel
#KERNEL_VERS=2.6.21.y
KERNEL_VERS=3.4.y
KERNEL_VERS_BRANCH=linux-${KERNEL_VERS}-ccm2200

# 2017-05-16 gc: uncomment this line to build kernel for jffs2 rootfs
UBIFS_CONFIG="-ubifs"

#prepare buildroot dl directory
(
  cd $BUILDROOT_BASE
#  tar xvzf $1/sources/userland/application_sources.tar.gz
#  mkdir dl
#  cd dl
  git clone https://github.com/swarco/swarco-linux-v2-dl
  rm -rf dl
  ln -s swarco-linux-v2-dl/dl dl
)

# 2011-01-07 gc: prepare a copy of buildroot 2 to create the soft-float
# toolchain necessary to build u-boot
(
  cd $BUILDROOT_BASE

  # if [ -d $BUILDROOT_DIR/.svn ]
  # then 
  #   svn export $BUILDROOT_DIR $BUILDROOT_SOFT_FLOAT_DIR
  # fi

  # fallback if svn export is not available (weiss-linux tree is 
  # created from .tar.gz)
  if ! [ -f $BUILDROOT_SOFT_FLOAT_DIR/Makefile ]
  then
    cp -aT $BUILDROOT_DIR $BUILDROOT_SOFT_FLOAT_DIR
  fi

  ( cd $BUILDROOT_SOFT_FLOAT_DIR; ./prepare_tree_soft_float.sh )
  cp dot_config_buildroot-2.0-soft-float $BUILDROOT_SOFT_FLOAT_DIR/.config  
)


#prepare buildroot 2.0 tree

(
    cd $BUILDROOT_BASE/buildroot-2.0
    ./prepare_tree.sh
)


#prepare ccm2200 u-boot directory
(
    cd $U_BOOT_BASE
    
    \rm -rf $U_BOOT_DIR
    git clone -b v2010.09-ccm2200 https://github.com/swarco/u-boot $U_BOOT_DIR
)


#prepare ccm2200 kernel directory
(
    KERNEL_OUTPUT_DIR=output-${KERNEL_VERS}-ccm2200
    KERNEL_CCM2200_DIR=linux-${KERNEL_VERS}-ccm2200
    KERNEL_BOARD_DIR=$PWD/board/swarco/ccm2200/kernel/${KERNEL_VERS}
    cd $KERNEL_BASE

    if ! [ -d $KERNEL_CCM2200_DIR ] ; then
        git clone -b ${KERNEL_VERS_BRANCH} https://github.com/swarco/linux-kernel $KERNEL_CCM2200_DIR
    else
        (
            cd $KERNEL_CCM2200_DIR
            git clean -d -x -f
            git checkout ${KERNEL_VERS_BRANCH}
        )
    fi
    mkdir $KERNEL_OUTPUT_DIR
    cp $KERNEL_BOARD_DIR/kernel-config-${KERNEL_CCM2200_DIR##*linux-}$UBIFS_CONFIG $KERNEL_OUTPUT_DIR/.config
    (
            cd ${KERNEL_CCM2200_DIR}
            ln -sf ../build-ccm2200.sh .
    )
)
echo KERNEL_VERS=${KERNEL_VERS} > version.mk

# Local Variables:
# mode: shell-script
# compile-command: "./prepare_tree_github.sh"
# End:

